contracts/zero-ex/contracts/src/features/OtcOrdersFeature.sol

```diff
@@ -0,0 +1,334 @@
+    /// @dev Version of this feature.
+    uint256 public immutable override FEATURE_VERSION = _encodeVersion(1, 0, 0);
+
+    constructor(address zeroExAddress)
+        public
+        FixinEIP712(zeroExAddress)
+    {
+        // solhint-disable no-empty-blocks
+    }
+
+    /// @dev Initialize and register this feature.
+    ///      Should be delegatecalled by `Migrate.migrate()`.
+    /// @return success `LibMigrate.SUCCESS` on success.
+    function migrate()
+        external
+        returns (bytes4 success)
+    {
+        _registerFeatureFunction(this.fillOtcOrder.selector);
+        _registerFeatureFunction(this.fillTakerSignedOtcOrder.selector);
+        _registerFeatureFunction(this.getOtcOrderInfo.selector);
+        _registerFeatureFunction(this.getOtcOrderHash.selector);
+        return LibMigrate.MIGRATE_SUCCESS;
+    }
+
+    /// @dev Fill a OTC order for up to `takerTokenFillAmount` taker tokens.
+    /// @param order The OTC order.
+    /// @param makerSignature The order signature from the maker.
+    /// @param takerTokenFillAmount Maximum taker token amount to fill this
+    ///        order with.
+    /// @return takerTokenFilledAmount How much maker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function fillOtcOrder(
+        LibNativeOrder.OtcOrder memory order,
+        LibSignature.Signature memory makerSignature,
+        uint128 takerTokenFillAmount
+    )
+        public
+        override
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        if (msg.sender != order.taker) {
+            bytes32 orderHash = getOtcOrderHash(order);
+            LibNativeOrdersRichErrors.OrderNotFillableByTakerError(
+                orderHash,
+                msg.sender,
+                order.taker
+            ).rrevert();
+        }
+        LibSignature.Signature memory nullSignature;
+        return _fillOtcOrderPrivate(
+            order,
+            makerSignature,
+            nullSignature,
+            takerTokenFillAmount
+        );
+    }
+
+    /// @dev Fill a OTC order for up to `takerTokenFillAmount` taker tokens.
+    ///      "Meta-transaction" variant, requires order to be signed by both
+    ///      maker and taker.
+    /// @param order The OTC order.
+    /// @param makerSignature The order signature from the maker.
+    /// @param takerSignature The order signature from the taker.
+    /// @param takerTokenFillAmount Maximum taker token amount to fill this
+    ///        order with.
+    /// @return takerTokenFilledAmount How much maker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function fillTakerSignedOtcOrder(
+        LibNativeOrder.OtcOrder memory order,
+        LibSignature.Signature memory makerSignature,
+        LibSignature.Signature memory takerSignature,
+        uint128 takerTokenFillAmount
```

Comment:
Partial fills here are tricky. The taker then isn't guaranteed any particular amount by signing, the `txOrigin` completely dictates the amount.

Reply:
Not a security flaw but I think worth discussing.

Reply:
discussed offline; decided to remove partial fills for this function